r/src/arrow_exports.h

```diff
@@ -0,0 +1,56 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// This gets included in arrowExports.cpp
+
+#pragma once
+
+#include "arrow_rcpp.h"
+
+#if defined(ARROW_R_WITH_ARROW)
+#include <arrow/csv/reader.h>
+#include <arrow/dataset/api.h>
```

Comment:
I suspect if there were more type_fwd.h files (like for dataset), we could trim this file back further and shave off some more seconds.

Reply:
It ain't going to help, you'll have to include `<arrow/dataset/api.h` in `arrowExports.cpp` anyway (since it calls the dataset constructor/destructor/move.

Reply:
Are you sure? IIUC (which I may very well not) arrowExports.cpp itself doesn't call the constructors etc., it just wraps, so you just need to have the types in the signature declared somewhere. See e.g. https://github.com/apache/arrow/blob/master/r/src/arrowExports.cpp#L1534-L1547